bash.sh

```diff
@@ -12,40 +12,83 @@
 #
 # The only proper way to do this is by looping through each item. Other
 # solutions claim to do this using string matching, but they are often
-# dangerous because don't handle edge cases.
+# dangerous because they don't handle edge cases, e.g. may be susceptible to
+# partial match or may not properly handle elements containing newlines.
 #
 
 #
 # Returns 0 if item is in the array; 1 otherwise.
 #
-# $1: The array value to search for
-# $@: The array values, e.g. "${myarray[@]}"
+# $1: name of the array variable to search
+# $2: array item to search for
 #
 array_contains() {
-    local item=$1; shift
-    for val in "$@"; do
+    local arr_name=$1
+    local item=$2
+
+    eval "local -a _tmp_arr=(\"\${${arr_name}[@]}\")"
+
+    for val in "${_tmp_arr[@]}"; do
         if [ "$val" == "$item" ]; then
             return 0
         fi
     done
+
+    return 1
+}
+
+
+#
+# Returns 0 if regexp matches any item in the array; 1 otherwise.
+#
+# $1: name of the array variable to search
+# $2: regexp to match
+#
+array_contains_regexp() {
+    local arr_name=$1
+    local item=$2
+
+    eval "local -a _tmp_arr=(\"\${${arr_name}[@]}\")"
+
+    for val in "${_tmp_arr[@]}"; do
+        if [[ "$val" =~ $item ]]; then
+            return 0
+        fi
+    done
+
     return 1
 }
 
 # ### Array filter
 
 #
-# Return all elements of an array with the specified item removed.
+# Remove elements from the named array. Iterate over each element and call
+# named function. If the function returns false, the element is removed.
 #
-# $1: The array value to remove
-# $@: The array values, e.g. "${myarray[@]}"
+# $1: name of the array variable
+# $@: name of the function to call plus any arguments; element to test will be
+#     provided as the last arg
 #
 array_filter() {
-    local item=$1; shift
-    for val in "$@"; do
-        if [ "$val" != "$item" ]; then
-            echo $val
+    local arr_name=$1; shift
+    local func_name=$1; shift
+    local -a func_args=("$@")
+
+    # Escape func args
```

Comment:
random question: should this be a separate function? or do you want to have them independent? I'd personally be fine with a re-use, I think.

Also, I don't understand why we need escaping, I thought `${arr_name}[@]` will correctly contain elements we need?

```bash
pawel@pawel-C02V306VHTDH ~ $ cat test.sh
#!/bin/bash

set -eu

func() {
    echo $#
    echo $@
}

func 123 123 123
pawel@pawel-C02V306VHTDH ~ $ bash test.sh
3
123 123 123
```

Note: general sense: you could also be OVERVERBOSE in your comments and explain why exactly things are happening?